File: 2021/02-drive/03.mjs
import { promises as fs } from "node:fs"

const parseCommand = str => {
    const [command, amount] = str.split(" ")
    return { command, amount: parseInt(amount) }
}

const applyCommand = ({ pos: [x, y], aim }, { command, amount }) => {
    switch (command) {
    case "forward":
        return { pos: [x + amount, y + aim * amount], aim }
    case "down":
        return { pos: [x, y], aim: aim + amount }
    case "up":
        return { pos: [x, y], aim: aim - amount }
    }
}

const file = await fs.readFile("input.txt")
const { pos: [x, y] } = file
    .toString()
    .trim()
    .split("\n")
    .map(parseCommand)
    .reduce(applyCommand, { pos: [0, 0], aim: 0 })

console.log(x * y)

File: 2021/02-drive/03.py
file = open("input.txt")
contents = file.read().strip()
file.close()

x = 0
y = 0
aim = 0
for line in contents.split("\n"):
    command, amount = line.split(" ")
    if command == "forward":
        x += int(amount)
        y += aim * int(amount)
    elif command == "down":
        aim += int(amount)
    elif command == "up":
        aim -= int(amount)

print(x * y)
